src/plugins/links/public/components/editor/links_editor.tsx

```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+
+import {
+  EuiForm,
+  EuiBadge,
+  EuiTitle,
+  EuiButton,
+  EuiSwitch,
+  EuiFormRow,
+  EuiToolTip,
+  EuiFlexItem,
+  EuiFlexGroup,
+  EuiDroppable,
+  EuiDraggable,
+  EuiFlyoutBody,
+  EuiButtonEmpty,
+  EuiButtonGroup,
+  EuiFlyoutFooter,
+  EuiFlyoutHeader,
+  EuiDragDropContext,
+  euiDragDropReorder,
+  EuiButtonGroupOptionProps,
+} from '@elastic/eui';
+import { DashboardContainer } from '@kbn/dashboard-plugin/public/dashboard_container';
+
+import { LinksLayoutInfo } from '../../embeddable/types';
+import {
+  Link,
+  LinksLayoutType,
+  LINKS_HORIZONTAL_LAYOUT,
+  LINKS_VERTICAL_LAYOUT,
+} from '../../../common/content_management';
+import { coreServices } from '../../services/kibana_services';
+import { LinksStrings } from '../links_strings';
+import { openLinkEditorFlyout } from '../../editor/open_link_editor_flyout';
+import { memoizedGetOrderedLinkList } from '../../editor/links_editor_tools';
+import { PanelEditorLink } from './panel_editor_link';
+import { LinksEditorEmptyPrompt } from './links_editor_empty_prompt';
+
+import { TooltipWrapper } from '../tooltip_wrapper';
+
+import './links_editor.scss';
+
+const layoutOptions: EuiButtonGroupOptionProps[] = [
+  {
+    id: LINKS_VERTICAL_LAYOUT,
+    label: LinksLayoutInfo[LINKS_VERTICAL_LAYOUT].displayName,
+  },
+  {
+    id: LINKS_HORIZONTAL_LAYOUT,
+    label: LinksLayoutInfo[LINKS_HORIZONTAL_LAYOUT].displayName,
+  },
+];
+
+const LinksEditor = ({
+  onSaveToLibrary,
+  onAddToDashboard,
+  onClose,
+  initialLinks,
+  initialLayout,
+  parentDashboard,
+  isByReference,
+}: {
+  onSaveToLibrary: (newLinks: Link[], newLayout: LinksLayoutType) => Promise<void>;
+  onAddToDashboard: (newLinks: Link[], newLayout: LinksLayoutType) => void;
+  onClose: () => void;
+  initialLinks?: Link[];
+  initialLayout?: LinksLayoutType;
+  parentDashboard?: DashboardContainer;
+  isByReference: boolean;
+}) => {
+  const toasts = coreServices.notifications.toasts;
+  const editLinkFlyoutRef: React.RefObject<HTMLDivElement> = useMemo(() => React.createRef(), []);
+
+  const [currentLayout, setCurrentLayout] = useState<LinksLayoutType>(
+    initialLayout ?? LINKS_VERTICAL_LAYOUT
+  );
+  const [isSaving, setIsSaving] = useState(false);
+  const [orderedLinks, setOrderedLinks] = useState<Link[]>([]);
+  const [saveByReference, setSaveByReference] = useState(!initialLinks ? true : isByReference);
```

Comment:
I wonder if this is intentional that we save to library by default. I thought that we usually try to save by value by default.

Reply:
This is intentional. 

Links are expected to be used to navigate between several dashboards. So it makes sense for links to be by reference by default. This way someone can add another link to a Links panel on one dashboard that automatically applies to all other linked dashboards.

If the Links panel is by-value, then each panel must be updated separately on all dashboards. We'd expect users don't want to have to do that.